usb: device_next: usb3 support

josuah opened the pull request:
This adds definitions needed for USB3 support.

There is no USB3 controller supported upstream yet. Should I wait to have the driver contributed in the same PR?

Here is the result of USB3CV for a this upcoming driver:
https://josuah.net/paste/rOzDHBjCbAIWO8d4qZu3/Chapter_9_Tests_USB_3_Gen_X_Failed_2024_12_10_08_18_16.html

It fails at SetFeature, as well as re-enumerating as USB2 (but this is related to the driver, not the USB3 handling).
But everything else is passing.

It also runs well with Twister on SuperSpeed configuration:

<details><summary>
<code>west twister --platform native_sim --scenario tests/subsys/usb/device_next/usb.device_next --inline-logs</code>
</summary>
<pre>Renaming output directory to /home/tinyvision/zephyrproject/zephyr/twister-out.1
INFO    - Using Ninja..
INFO    - Zephyr version: v4.0.0-1781-gbb86ae133c77
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/tinyvision/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 1
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:    1/   1  100%  built (not run):    0, filtered:    0, failed:    0, error:    0
INFO    - 1 test scenarios (1 configurations) selected, 0 configurations filtered (0 by static filter, 0 at runtime).
INFO    - 1 of 1 executed test configurations passed (100.00%), 0 built (not run), 0 failed, 0 errored, with no warnings in 25.02 seconds.
INFO    - 2 of 2 executed test cases passed (100.00%) on 1 out of total 880 platforms (0.11%).
INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/tinyvision/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/tinyvision/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/tinyvision/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed</pre>
</details>

josuah commented:
force-push:
- CI fixes
- doc fixes

XenuIsWatching commented on include/zephyr/drivers/usb/udc.h:
```diff
@@ -36,6 +37,8 @@ enum udc_mps0 {
 struct udc_device_caps {
 	/** USB high speed capable controller */
 	uint32_t hs : 1;
+	/** USB high speed capable controller */
```
Copy and paste typo here. Should be "super" rather than "high"

josuah replied:
Good catch, thank you!

josuah commented:
force-push:
- Fix a typo for the udc caps struct
- Add `SET_ISOCH_DELAY` as a no-op, as the host informs the device of an end-to-end delay observed, not used at this stage, but it might help with compliance to support it even if it does nothing.

josuah commented:
The USB3 driver in question is soon to be contributed, and the implementation lives here in the meantime:
- https://github.com/tinyvision-ai-inc/tinyvision_zephyr_sdk/tree/main/drivers/usb/udc

tmon-nordic commented on include/zephyr/usb/bos.h:
```diff
@@ -41,6 +42,13 @@ struct usb_bos_capability_lpm {
 	uint32_t bmAttributes;
 } __packed;
 
+/** Fields for @ref usb_bos_capability_lpm bmAttributes */
```
LPM is for USB 2.0 EXTENSION capability descriptor and LTM is for SUPERSPEED USB capability descriptor. Please do not mix different `bmAttributes` bits in single enum.

josuah replied:
I adjusted this in the latest commits, thank you!

tmon-nordic commented on include/zephyr/drivers/usb/udc.h:
```diff
@@ -62,6 +65,20 @@ enum udc_bus_speed {
+ * device can configure the hardware to match what is on the host.
+ * That way, both side of the session have the same properties.
+ */
+struct udc_exit_latency {
```
The specification refers to these as "System Exit Latency". `udc_exit_latency` seems to suggests it is some software construct and not a standard requirement. Not sure what would be a better name though.

josuah replied:
Good point. It is a packed struct defined by the standard, so can go to the USB headers rather than UDC ones.
![scrot_20250111_135633_1422x684](https://github.com/user-attachments/assets/7d815842-240a-47a2-b98b-bc4d4d73765d)

tmon-nordic commented on include/zephyr/usb/usb_ch9.h:
```diff
@@ -105,6 +105,20 @@ static inline bool usb_reqtype_is_to_device(const struct usb_setup_packet *setup
+#define USB_SREQ_SET_WUSB_DATA		0x14
+#define USB_SREQ_LOOPBACK_DATA_WRITE	0x15
+#define USB_SREQ_LOOPBACK_DATA_READ	0x16
+#define USB_SREQ_SET_INTERFACE_DS	0x17
```
What about `GET_FW_STATUS` and `SET_FW_STATUS`?

josuah replied:
I had no idea this existed!
https://usbpd.wordpress.com/wp-content/uploads/2019/02/25-usb_devdays_taipei_2017_cto_hour.pdf

![scrot_20250111_183400_1288x720](https://github.com/user-attachments/assets/2b865443-213c-4dd8-a8c6-3edf95206ee8)

josuah replied:
I found something for the link layers: [[UCSI](https://www.usb.org/sites/default/files/USB%20Type-C%20Connector%20System%20Software%20Interface%20UCSI%20Revision%202.0%20-%20FINAL.pdf)]
![scrot_20250111_185848_1042x284](https://github.com/user-attachments/assets/4dd56650-67d0-4de5-8c12-f152101286d3)

I found something for the Type-C Power Delivery (alternative to DFU) [[PDFU](https://usb.org/sites/default/files/documents/pd_fw_update_specification_rev_1_0_20160915b.pdf)]
![scrot_20250111_190538_1282x465](https://github.com/user-attachments/assets/689f05c1-6f93-46cb-91bd-1d1e65d88480)

But nothing about `GET_FW_UPDATE` itself... Any hint?

tmon-nordic replied:
Get USB 3.2 Revision 1.1 Specification from https://www.usb.org/document-library/usb-32-revision-11-june-2022

josuah replied:
I had the 2017 version somehow. Thank you!

tmon-nordic commented on include/zephyr/usb/usb_ch9.h:
```diff
@@ -121,6 +135,7 @@ static inline bool usb_reqtype_is_to_device(const struct usb_setup_packet *setup
 #define USB_DESC_INTERFACE_ASSOC	11
 #define USB_DESC_BOS			15
 #define USB_DESC_DEVICE_CAPABILITY	16
+#define USB_DESC_ENDPOINT_COMPANION	48
```
There is also `SUPERSPEEDPLUS_ISOCHRONOUS_ENDPOINT_COMPANION`

josuah replied:
That makes sense, SuperSpeed Plus using an RTOS are [already on the market](https://www.eenewseurope.com/en/first-usb-3-2-gen2-superspeed-camera/) (but not available yet).

I might also need to add the structs for it then...

tmon-nordic commented on include/zephyr/usb/usb_ch9.h:
```diff
@@ -248,6 +263,15 @@ struct usb_association_descriptor {
 } __packed;
 
+/** USB Endpoint Companion Descriptor defined in USB3 spec. Table 9-27. */
+struct usb_co_descriptor {
```
I would go with `usb_ss_endpoint_companion_descriptor`.

josuah replied:
I did not dare to break the regularity of every class.

https://github.com/zephyrproject-rtos/zephyr/blob/5750b720b37a8756a8578eae319a9c0e3f0fe1df/subsys/usb/device_next/class/loopback.c#L34-L64

Should that be `usb_ss_ep_companion_descriptor` for consistency with `usb_ep_descriptor`?

I am applying `usb_ss_endpoint_companion_descriptor` currently...

josuah replied:
I also renamed `if..._ss_..._co` to `if..._ss_.._ep_co`.

tmon-nordic reviewed (commented):
The fact that Super-Speed requires bulk endpoints wMaxPacketSize to be 1024 makes https://github.com/zephyrproject-rtos/zephyr/pull/76255 even more important

josuah commented:
force-push:
- Renamed `udc_exit_latency` into `usb_system_exit_latency` and moved it to `usb_ch9.h`
- Added USB3 SuperSpeedPlus capability descriptor (untested)
- Added `USB_BOS_CAPABILITY_PRECISION_TIME_MEASUREMENT` 
- Renamed and moved capabilities as requested.
- Wording, and section number typo fixups

SuperSpeedPlus still lacks other modifications to be enabled: everywhere SuperSpeed (SS) is introduced, SuperSpeedPlus (SSP) needs to be introduced as well.

josuah commented:
> The fact that Super-Speed requires bulk endpoints wMaxPacketSize to be 1024 makes #76255 even more important

Here is what this PR it looks like on top of #76255:
https://github.com/tmon-nordic/zephyr/compare/max-speed...tinyvision-ai-inc:zephyr:pr-usb3-max-speed

I thought it would be more suited to wait #76255 to be rebased before force-pushing this here.
Thank you for this review, and thanks all for this long-term maintenance effort of the new USB stack!

josuah commented:
force-push:
- add changes from [2022 version of the standard](https://www.usb.org/document-library/usb-32-revision-11-june-2022) (R1.1): `SET_FW_UPDATE` and `GET_FW_UPDATE`
- doc fix

josuah commented:
force-push:

- apply changes from https://github.com/zephyrproject-rtos/zephyr/pull/76255
- rebase on latest `main`

josuah commented:
> Super-Speed Gen1x1/Gen1x2/Gen2x1/Gen2x2 naming

It might be possible to dodge the problem by using:

- `USBD_SUPPORTS_SUPER_SPEED` (Gen1x1, Gen1x2) introduced now (DWC3, FX5),
- `USBD_SUPPORTS_SUPER_SPEED_PLUS` (Gen2x1, Gen2x2) added later (FX10/FX20),

Combined with this to describe the full Gen1x1, Gen1x2, Gen2x1, Gen2x2:

- `USBD_SUPPORTS_DUAL_LANE` (Gen1x2, Gen2x2) added later (FX5N, FX20).

josuah committed:
usb: device_next: USB 3 support

Introduce definitions to enable USB3 device controllers to be written.
So far, USB 3 support got added to the subsystem, UDC common code, the
Loopback class, UVC class, and the CDC-ACM class.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
Force-push:
- Fix CDC ACM which was not completely working anymore after the latest rebase
